### PR TITLE
Patch for CLIENTID command in SMTP and IMAP Low-level API

### PR DESCRIPTION
**Pull request by shaun-a-johnson**
This is a patch to the 'libetpan' library, to support new CAPABILITY strings, a security extension for traditional SMTP and IMAP protocols.

This patch to the low-level SMTP and IMAP API will expose support for the ability to recognize the new capability, and present a new command called CLIENTID to be presented to the server.

There's RFC drafts currently in the works for both the SMTP and IMAP "CLIENTID" commands:
  SMTP - https://tools.ietf.org/html/draft-storey-smtp-client-id-05
  IMAP - https://tools.ietf.org/html/draft-yu-imap-client-id-00

Allowing this patch to become part of the main libetpan library will enable us to complete CLIENTID patches for several email clients which depend on the libetpan library for their SMTP and IMAP support.

That development work will allow those email clients to leverage CLIENTID security available in mail platforms that support this added feature.

This will not have any impact on legacy behavior.

### CODE REVIEW

**Comment by dinhviethoa** [comment]
Thanks for adding this feature!
Could you move IMAP CLIENTID to a file named mailimap_clientid.c & friends?
Also, is there a way to detect this capability on IMAP?
I think the location is fine for SMTP.

**Comment by linuxmagic-mp** [comment]
Yes, per the RFC Draft, CLIENTID is advertised in the IMAP capability string, however it normally only would be advertised ONLY if the connection was 'secured', eg SSL/TLS.
eg.. openssl s_client -connect <FQDN>:993
* OK [CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE CLIENTID AUTH=PLAIN] MagicMail ready.
When looking at those changes, will look at making that more clear in documentation

**Comment by dinhviethoa** [comment]
Could you add a function similar to this one?
https://github.com/dinhviethoa/libetpan/blob/master/src/low-level/imap/namespace.h#L51

**Comment by shaun-a-johnson** [comment]
> Could you add a function similar to this one?
> https://github.com/dinhviethoa/libetpan/blob/master/src/low-level/imap/namespace.h#L51
Just to clarify, are you only asking for the capability detection function similar to mailimap_has_namespace? 

Or are you asking for clientid to be converted to the 'extension' format shown in namespace.h?

> Could you move IMAP CLIENTID to a file named mailimap_clientid.c & friends?
What about the mailimap_clientid_send() function in mailimap_sender.c, would you like that to stay there?

**Comment by dinhviethoa** [comment]
I'm asking two things:
- Move the functions out of the main files
- Add a capability detection function similar to mailimap_has_namespace
(I think we don't need to do something as complex as namespace here)

Please move mailimap_clientid_send() to clientid_sender.[ch].

**Comment by shaun-a-johnson** [comment]
Hello and thanks again for taking the time to review this pull request, I believe I have made the changes you asked for. When you have time please look over this new set of changes and let me know if there is anything I have done incorrectly or anything I can improve.

File: src/low-level/smtp/mailsmtp_types.h [hunk 1]
```diff
@@ -70,6 +70,7 @@ enum {
   MAILSMTP_ERROR_AUTH_ENCRYPTION_REQUIRED,
   MAILSMTP_ERROR_STARTTLS_TEMPORARY_FAILURE,
   MAILSMTP_ERROR_STARTTLS_NOT_SUPPORTED,
+  MAILSMTP_ERROR_CLIENTID_NOT_SUPPORTED,
```
Comment (dinhviethoa) [inline src/low-level/smtp/mailsmtp_types.h]: Could you move it to the end of the emum?

**Comment by dinhviethoa** [comment]
One last thing then it should be all good!

File: src/low-level/smtp/mailsmtp_types.h (hunk 1 above)
Comment (shaun-a-johnson) [reply src/low-level/smtp/mailsmtp_types.h]: Would you also like the IMAP error in mailimap_types.h moved?

> MAILIMAP_ERROR_CLIENTID

**Comment by shaun-a-johnson** [comment]
Would you also like the error `MAILIMAP_ERROR_CLIENTID` in mailimap_types.h moved to the end of the enum?

**Comment by dinhviethoa** [comment]
> Would you also like the error MAILIMAP_ERROR_CLIENTID in mailimap_types.h moved to the end of the enum?

Missed this one! Yes. Please!

**Comment by shaun-a-johnson** [comment]
Alright, thanks again. I think I have wrapped up everything in this last push.  
Give it one last look over and let me know if there is anything I have missed.

**Comment by dinhviethoa** [comment]
Thank you! Merged.

**Pull request merged**